lib/core/include/scipp/core/spatial_transforms.h

```diff
@@ -0,0 +1,139 @@
+inline const auto &asEigenType(const Quaternion &obj) { return obj.quat(); }
+inline auto asEigenType(const Translation &obj) {
+  return Eigen::Translation<double, 3>(obj.vector());
+}
+
+template <class T_LHS, class T_RHS>
+struct combines_to_linear : std::false_type {};
+template <>
+struct combines_to_linear<Quaternion, Eigen::Matrix3d> : std::true_type {};
+template <>
+struct combines_to_linear<Eigen::Matrix3d, Quaternion> : std::true_type {};
+
+template <typename T_LHS, typename T_RHS>
+[[nodiscard]] inline std::enable_if_t<combines_to_linear<T_LHS, T_RHS>::value,
+                                      Eigen::Matrix3d>
+operator*(const T_LHS &lhs, const T_RHS &rhs) {
+  return asEigenType(lhs) * asEigenType(rhs);
+}
+
+template <class T_LHS, class T_RHS>
+struct combines_to_affine : std::false_type {};
+template <>
+struct combines_to_affine<Quaternion, Translation> : std::true_type {};
+template <>
+struct combines_to_affine<Quaternion, Eigen::Affine3d> : std::true_type {};
+template <>
+struct combines_to_affine<Eigen::Matrix3d, Translation> : std::true_type {};
+template <>
+struct combines_to_affine<Eigen::Affine3d, Quaternion> : std::true_type {};
+template <>
+struct combines_to_affine<Eigen::Affine3d, Translation> : std::true_type {};
+template <>
+struct combines_to_affine<Translation, Quaternion> : std::true_type {};
+template <>
+struct combines_to_affine<Translation, Eigen::Matrix3d> : std::true_type {};
+template <>
+struct combines_to_affine<Translation, Eigen::Affine3d> : std::true_type {};
+
+template <typename T_LHS, typename T_RHS>
+[[nodiscard]] inline std::enable_if_t<combines_to_affine<T_LHS, T_RHS>::value,
+                                      Eigen::Affine3d>
+operator*(const T_LHS &lhs, const T_RHS &rhs) {
+  return Eigen::Affine3d(asEigenType(lhs) * asEigenType(rhs));
+}
+
+template <class T_LHS> struct applies_to_vector : std::false_type {};
+template <> struct applies_to_vector<Quaternion> : std::true_type {};
+template <> struct applies_to_vector<Translation> : std::true_type {};
+
+template <typename T_LHS>
+[[nodiscard]] inline std::enable_if_t<applies_to_vector<T_LHS>::value,
+                                      Eigen::Vector3d>
+operator*(const T_LHS &lhs, const Eigen::Vector3d &rhs) {
+  return asEigenType(lhs) * rhs;
+}
+
+[[nodiscard]] inline Quaternion operator*(const Quaternion &lhs,
+                                          const Quaternion &rhs) {
+  return Quaternion(lhs.quat() * rhs.quat());
+};
+
+[[nodiscard]] inline Translation operator*(const Translation &lhs,
+                                           const Translation &rhs) {
+  return Translation(lhs.vector() + rhs.vector());
+};
+
+template <> inline constexpr DType dtype<Eigen::Matrix3d>{5001};
+template <> inline constexpr DType dtype<Eigen::Affine3d>{5002};
+template <> inline constexpr DType dtype<Translation>{5003};
+template <> inline constexpr DType dtype<Quaternion>{5004};
```

Comment:
Confusing that this was split up in this way. Just put all the spatial transforms and eigen types starting at 4000?